### PR TITLE
Redshift data sharing - Boilerplate for redshift dataset sharing module

### PR DESCRIPTION
**Pull request by dlpzx**
### Feature or Bugfix
- Feature

### Detail
Complete design in https://github.com/data-dot-all/dataall/issues/955. This particular PR is part of the redshift dataset sharing implementation. 
- it adds the `redshift_datasets_shares` module package and its __init__ file
- it creates the main subpackages in the module
- it registers the share processor and creates the base share processor
- in db.repository it adds a EnvironmentResource class to prevent orphan resources left in environment deletes
- it adds the module in config.json
- It adds RedshiftTable as a ShareableItem
- It adds RedshiftRole as a PrincipalType (also remove unused types, including one only used in testing the subscription task)

### Relates
- #955 

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dlpzx** [comment]
With this changes we can go through the normal share request workflow for redshift tables. The sharing when approved does not work; that will be done in a separate PR
https://github.com/user-attachments/assets/e131c86a-3502-47de-8352-6eb8ee08dd3f

File: backend/dataall/modules/redshift_datasets_shares/db/redshift_share_object_repositories.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            .filter(
+                and_(
+                    ShareObject.environmentUri == environment_uri,
+                    ShareObject.groupUri == group_uri,
+                    ShareObject.principalType == principal_type.value,
+                    ShareObjectItem.itemType == ShareableType.RedshiftTable.value,
+                )
+            )
+            .count()
+        )
```
[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/redshift_datasets_shares/db/redshift_share_object_repositories.py:47]: Let me give it a second round, I followed what it was implemented in s3_datasets assuming it was right. 
--> in this case we do not need a count_role_resources; consumption roles cannot be the principal of a share request, so we do not need to check if there are shares for a consumption role
--> for deletes, the behavior should be: 1) count resources counts the SHARED objects, 2) delete deletes all NON-SHARED (draft, pending...)

[the remark this answers is not in the document]
Comment (noah-paige) [reply backend/dataall/modules/redshift_datasets_shares/db/redshift_share_object_repositories.py]: yes I see the changes you made and I think looks good for count and delete of redshift shares

can open a separate PR to fix any issues with S3 shares

**Review by noah-paige — APPROVED** [review]
lgtm - approving

File: backend/dataall/modules/redshift_datasets_shares/db/redshift_share_object_repositories.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/redshift_datasets_shares/db/redshift_share_object_repositories.py]: Sounds good

**Pull request merged**